documentation/correctness.adoc

```diff
@@ -0,0 +1,97 @@
+
+Whether a model is _correct_ –- or _valid_, or _consistent_ –- is an important question.
+Correctness ultimately determines whether a model is useful in the context it exists in.
+Answering that question only with a simple "`yes`" or "`no`" is too simplistic: a model can be incorrect in different ways and on different levels.
+
+Not being able to resolve a reference is a different kind of failure than violating a semantic constraint.
+Having an illegal extra comma in an array somewhere in the JSON serialization chunk of a model -- e.g. `[ 0, 1, 2**,** ]` -- should not prevent that serialization chunk from being deserializable.
+
+[IMPORTANT]
+====
+We'll talk about _correctness_ from now, and consider _validity_ and _consistency_ to be equivalent concepts.
+====
+
+It makes sense to phrase the answer as a collection of _issues_, instead of giving a hard "`yes/no`" answer.
+An issue is a specific point in the model where a specific kind of incorrectness arises.
+Many (kinds of) issues can be meaningfully recovered from, and recovery can often be obvious to the point it can be automatic.
+The unresolved reference can be left in as-is, and treated as an unknown (but not a _missing_) value with special meaning (i.e., domain-specific semantics).
+The illegal comma should be reported but can subsequently be safely ignored.
+In cases where recovery is not automatic, an issue should be very clear to the user what they can do to resolve it.
+A model must be editable/mutable -- and even usable to some -- even if it has issues.
+
+LionWeb provides a precise notion for model correctness in the form of a multi-leveled specification.
+For each level, it specifies what kinds of incorrectness can arise, how these should be reported as issues, and whether and how these can be meaningfully _recovered_ from.
+This approach provides a degree of resilience: issues arising in one level don't necessarily prohibit determining whether a model exhibits incorrectness on another level, provided an issue can be meaningfully _recovered_ from.
+
+LionWeb provides a JSON serialization format for models: the serialization chunk.
+LionWeb intentionally doesn't prescribe anything about the _programmatic representation_ of models -- i.e.: the runtime or in-memory representation that's the result of deserializing a serialization chunk of a model -- leaving implementors of language-oriented tooling free to choose the representation that's right for them.
+Because of that, this specification is phrased _completely_ in terms of the JSON serialization.
+The first three levels of this specification pertain purely to the JSON that's supposed to be the serialization chunk of a model, and are language(s)-aspecific.
+The next levels _could_ be phrased in terms of a programmatic representation, but the equivalent formulation purely in terms of the serialization format works just as well.
+
+This specification admits some flexibility on the first three, JSON-centric levels.
```

Comment:
Maybe we could borrow the term "well-formed" from XML for these three levels?

Reply:
Or maybe that would be "Hierarchical"?

Reply:
“Hierarchical” really pertains to inter-node (implied) references only (so level/aspect 3). I'll add a sentence mentioning “well-formed”.